parser/src/function.rs

```diff
@@ -14,76 +14,58 @@ pub(crate) struct ArgumentList {
     pub keywords: Vec<ast::Keyword>,
 }
 
-type ParameterDefs = (Vec<ast::Arg>, Vec<ast::Arg>, Vec<ast::Expr>);
-type ParameterDef = (ast::Arg, Option<ast::Expr>);
-
 // Perform validation of function/lambda arguments in a function definition.
-pub(crate) fn validate_arguments(
-    arguments: ast::Arguments,
-) -> Result<ast::Arguments, LexicalError> {
-    let mut all_args: Vec<&ast::Arg> = vec![];
-
-    all_args.extend(arguments.posonlyargs.iter());
-    all_args.extend(arguments.args.iter());
-
-    if let Some(a) = &arguments.vararg {
-        all_args.push(a);
-    }
-
-    all_args.extend(arguments.kwonlyargs.iter());
-
-    if let Some(a) = &arguments.kwarg {
-        all_args.push(a);
-    }
-
+pub(crate) fn validate_arguments(arguments: &ast::Arguments) -> Result<(), LexicalError> {
     let mut all_arg_names = FxHashSet::with_hasher(Default::default());
-    for arg in all_args {
-        let arg_name = &arg.arg;
-        // Check for duplicate arguments in the function definition.
+
+    for (range, arg_name) in arguments
+        .posonlyargs
+        .iter()
+        .chain(arguments.args.iter())
+        .chain(arguments.kwonlyargs.iter())
+        .map(|arg| (arg.def.range, arg.def.arg.as_str()))
+        .chain(
+            arguments
+                .vararg
+                .as_ref()
+                .map(|arg| (arg.range, arg.arg.as_str())),
+        )
+        .chain(
+            arguments
+                .kwarg
+                .as_ref()
+                .map(|arg| (arg.range, arg.arg.as_str())),
+        )
```

Comment:
Nit: This inline `in arguments...` is a lot to parse. Maybe assign the result to a temporary variable. This has the advantage that it gives the "thing" a name that guides readers.